src/traversals/diffusion.jl

```diff
@@ -0,0 +1,90 @@
+"""
+    diffusion(g, p, n)
+
+Run diffusion simulation on `g` for `n` steps with spread
+probabilities based on `p`. Return a vector with the set of vertices
+reached at each step of the simulation.
+
+### Optional Arguments
+- `initial_infections=sample(vertices(g), 1)`: A list of vertices that
+are infected at the start of the simulation.
+- `watch=Set()`: While simulation is always run on the full graph,
+specifying `watch` limits reporting to a specific set of vertices reached
+during the simulation. If left as an empty set, all vertices will be watched.
+- `normalize=false`: if `false`, set the probability of spread from a vertex ``i`` to
+each of the out_neighbors of ``i`` to ``p``. If `true`, set the probability of spread
+from a vertex ``i`` to each of the `out_neighbors` of ``i`` to
+``\\frac{p}{outdegreee(g, i)}``.
+"""
+
+function diffusion(g::AbstractGraph,
+                   p::AbstractFloat,
+                   n::Integer;
+                   watch::Vector=Vector{Int}[],
+                   initial_infections::Vector=LightGraphs.sample(vertices(g), 1),
+                   normalize::Bool=false
+                   )
+
+    # Initialize
+    T = eltype(g)
+    watch_set = Set{T}(watch)
+    infected_vertices = IntSet(initial_infections)
+    vertices_per_step::Vector{Vector{T}} = [Vector{T}() for i in 1:n]
+
+    # Record initial infection
+    if !isempty(watch_set)
+        watched_initial_infections = intersect(initial_infections, watch_set)
+        vertices_per_step[1] = T.(watched_initial_infections)
+    else
+        vertices_per_step[1] = T.(initial_infections)
+    end
+
+    # Run simulation
+    for step in 2:n
+        new_infections = Set{T}()
+
+        for i in infected_vertices
+            if normalize
+                local_p = convert(Float64, p / outdegree(g, i))::Float64
```

Comment:
I don't think the type assert or convert is necessary for local_p at all. If p is a rational and outdegree is an integer it should still be a rational. `randsubseq` should work with any input subtype of `Real`.

The outgree and out_neighbors is fine. its like 10 extra instructions max and this guarantees correct behavior in the more generic case.